src/lc3/vm.rs

```diff
@@ -137,4 +296,152 @@ mod tests {
+        let mut vm = VM::new();
+        vm.registers[1] = 1;
+        vm.add(instr)?;
+        assert_eq!(vm.registers[0], 0x3);
+        assert_eq!(vm.cond, ConditionFlag::Pos);
+        Ok(())
+    }
+
+    #[test]
+    fn and_register() -> Result<(), VMError> {
+        let instr: u16 = 0b0101_0000_0100_0010;
+        let mut vm = VM::new();
+        vm.registers[1] = 0x0F0F;
+        vm.registers[2] = 0xFFFF;
+        vm.and(instr)?;
+        assert_eq!(vm.registers[0], 0x0F0F);
+        assert_eq!(vm.cond, ConditionFlag::Pos);
+        Ok(())
+    }
+
+    #[test]
+    fn and_with_immediate() -> Result<(), VMError> {
+        let instr: u16 = 0b0001_0000_0110_0111;
+        let mut vm = VM::new();
+        vm.registers[1] = 0xFFFF;
+        vm.and(instr)?;
+        assert_eq!(vm.registers[0], 0b0111);
+        assert_eq!(vm.cond, ConditionFlag::Pos);
+        Ok(())
+    }
+
+    #[test]
+    fn not_register() -> Result<(), VMError> {
+        let instr: u16 = 0b1001_0000_0111_1111;
+        let mut vm = VM::new();
+        vm.registers[1] = 0x0F0F;
+        vm.not(instr)?;
+        assert_eq!(vm.registers[0], 0xF0F0);
+        assert_eq!(vm.cond, ConditionFlag::Neg);
+        Ok(())
+    }
+
+    #[test]
+    fn branch_on_flag() -> Result<(), VMError> {
+        let mut vm = VM::new();
+        vm.cond = ConditionFlag::Pos;
+        let instr: u16 = 0b0000_0010_0000_1010; // jump 10 places on positive flag
+        vm.br(instr)?;
+        assert_eq!(vm.pc, 0x300A);
+        Ok(())
+    }
+
+    #[test]
+    fn dont_branch_on_flag() -> Result<(), VMError> {
+        let mut vm = VM::new();
+        vm.cond = ConditionFlag::Neg;
+        let instr: u16 = 0b0000_0010_0000_1010; // jump 10 places on positive flag
+        vm.br(instr)?;
+        assert_eq!(vm.pc, 0x3000);
+        Ok(())
+    }
+
+    #[test]
+    fn jump_to_correct_value() -> Result<(), VMError> {
+        let mut vm = VM::new();
+        let instr: u16 = 0b1100_0000_0100_0000; // jump to value in R1
+        vm.registers[1] = 0x4242;
+        vm.jmp(instr)?;
+        assert_eq!(vm.pc, 0x4242);
+        Ok(())
+    }
+
+    #[test]
+    fn jump_to_subroutine() -> Result<(), VMError> {
+        let mut vm = VM::new();
+        let instr: u16 = 0b0100_1000_0100_0010; // add to PC 0x42
+        vm.jsr(instr)?;
+        assert_eq!(vm.pc, 0x3042);
+        Ok(())
+    }
+
+    #[test]
+    fn jump_to_subroutine_register() -> Result<(), VMError> {
+        let mut vm = VM::new();
+        let instr: u16 = 0b0100_0000_1100_0000; // set PC to R3 value
+        vm.registers[3] = 0x4242;
+        vm.jsr(instr)?;
+        assert_eq!(vm.pc, 0x4242);
+        Ok(())
+    }
+
+    #[test]
+    fn lea_store_address() -> Result<(), VMError> {
+        let mut vm = VM::new();
+        let instr: u16 = 0b1110_0000_0100_0010; // store in R0, PC plus 0x42
+        vm.lea(instr)?;
+        assert_eq!(vm.registers[0], 0x3042);
+        Ok(())
+    }
```

Comment:
Well done Tomi. Maybe, in a future PR, we could add tests that test operands working together rather than having a unit test for every operand.

Reply:
Nice, I think that would be great! 
but do you mean to replace the unit tests with that?

Reply:
It is not necessary to delete the unit test, but do tests than involve different operands, like perform and ADD instruction and then apply an AND instruction. When we have instruction related to memory this will be very helpful